seasonal ratings

catkins-miso opened the pull request:
- [x] "normal" proposal schema
- [x] "normal" proposal example
- [x] complete status response
- [x] incomplete status response
- [x] "slim" proposal schema
- [x] "slim" proposal example
- [x] detailed snapshot
- [x] "normal" snapshot

catkins-miso committed:
WIP

catkins-miso commented on docs/example-narratives/examples/seasonal-ratings-proposals-patch.json:
@getorymckeag there's a "slim" media type for this as well. This is the more general one. The main difference is this media type allows for arbitrary "seasons" to be defined per facility. In fact this is really just a forecast proposal.

catkins-miso commented on docs/example-narratives/examples/seasonal-ratings-proposals-slim-patch.json:
@getorymckeag this is the slim version. It works very much like the forthcoming "precompressed" forecast proposal.

catkins-miso commented on docs/example-narratives/examples/seasonal-ratings-proposals-slim-patch.json:
```diff
@@ -0,0 +1,99 @@
+          "begins": "2025-09-01T00:00:00-05:00"
+        }
+      ],
+      "ends": "2025-11-15T00:00:00-05:00"
```
@getorymckeag lmk what you think of this. I believe it addresses the RC-defined seasons use case but still maps to an explicit seasonal rating schedule. We'll need to write down a lot about how the PATCH is interpreted, but I think the server would require that the current date is included in one of the seasons and that at least a full year is covered from the `begins` of that current season through `ends`

catkins-miso commented on docs/example-narratives/examples/seasonal-ratings-proposals-slim-patch.json:
```diff
@@ -0,0 +1,99 @@
+      "ends": "2025-11-15T00:00:00-05:00"
+    },
+    "limit-type": {
+      "mva": 1
```
the value doesn't matter, but I wanted to re-use the limit-types schema. I'm totally open to more elegant ways to specify this. This will require some docs too, but if you had `{"mw":1, "pf":1}` here instead, then the array would just have two values instead of one. So `[normal_MVA, emergency_MVA]` becomes `[normal_MW, normal_pf, emerg_MW, emerg_pf]` e.g. `[50, 1.0, 55, 1.0]`. It is probably worth defining a new representation, but this is my first pass.

catkins-miso committed:
remove incorrect nesting

getorymckeag commented on docs/example-narratives/examples/seasonal-ratings-proposals-patch.json:
```diff
@@ -1,10 +1,52 @@
 {
+  "proposal-header": {
```
As we think about this here- do we need a separate set of examples for seasonal ratings for day vs night?

catkins-miso replied:
Yes, absolutely, thanks for reminding me.

catkins-miso committed:
clean up spectral errors

catkins-miso committed:
add day-night

catkins-miso commented on docs/example-narratives/examples/seasonal-ratings-proposals-slim-patch.json:
```diff
@@ -0,0 +1,73 @@
+          "begins": "2025-09-01T00:00:00-05:00"
+        }
+      ],
+      "ends": "2025-11-15T00:00:00-05:00"
```
Adding `"day-night":"DAY"` to the verbose format was straightforward, but everything I thought to do with the header in the slim format was awkward, so I opted to lean on the `PATCH` semantics and require a separate call for "DAY" and a separate call for "NIGHT". @getorymckeag 

I think that might require two GETs to read all the ratings assuming the PATCH and GET are symmetric. Not a fan of that.

I'll come up with something better tomorrow I hope.

catkins-miso committed:
ensure vscode extensions install

corp use of self-signed cert requires making node
aware outside of the postCreateCommand in order
to install extensions. The extensions pick this
up to so they can download things.

catkins-miso committed:
add description

catkins-miso committed:
tweak description

catkins-miso committed:
improve descriptions

catkins-miso committed:
not visible in redocly anyway

catkins-miso committed:
feature schema to have a way to show description

catkins-miso committed:
cleanup

catkins-miso committed:
refactor

catkins-miso committed:
update slim header description

clarifies the role of the names header

catkins-miso committed:
clean up

catkins-miso committed:
unused

catkins-miso committed:
add seasonal proposal status

catkins-miso committed:
add seasonal ratings snapshot

catkins-miso committed:
add media type for detailed seasonal ratings snapshot

catkins-miso committed:
ensure valid detailed examples

catkins-miso committed:
extract seasonal rating period from forecast file

catkins-miso committed:
update examples

catkins-miso commented on docs/_data/components/schemas/headers.yaml:
```diff
@@ -98,4 +95,52 @@ real-time-snapshot-header:
+      required:
+        - schedule
+        - ends
+    limit-type:
```
@getorymckeag Instead of doing this, let's parameterize the "slim" media type name, i.e., `application/vnd.trolie.seasonal-ratings-proposal-slim.v1+json; limit-type=apparent-power`

This leverages the media type standard to be explicit about the content without introducing a "slim" media type for each limit type. It also avoids the awkward explanations about why this requires a value that contributes nothing semantically.

The standard could define `Content-Type: application/json` in the `PATCH` to be interpreted at runtime by the Clearinghouse Provider as the "slim" media type with a fixed `limit-type=` parameter. The response would always have an explicit media type, of course, even when the client does a `GET Accept: */*, application/json`.

getorymckeag replied:
Sure looks good.  It's elegant but exotic, so I felt like I had to do a test on this- support for parameterized media types was pretty broken in Spring Web MVC for a long time.  I tried in the latest version and it appears to work however.

catkins-miso committed:
move limit-type to media type parameter
